ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java

```diff
@@ -606,6 +607,33 @@ private void handleAdminRequest(Request request) throws InterruptedException, IO
             }
           }
           break;
+        case ReplicationControl:
+          error = ServerErrorCode.No_Error;
+          metrics.replicationControlRequestQueueTimeInMs.update(requestQueueTime);
+          metrics.replicationControlRequestRate.mark();
+          processingTimeHistogram = metrics.replicationControlResponseQueueTimeInMs;
+          responseQueueTimeHistogram = metrics.replicationControlResponseQueueTimeInMs;
+          responseSendTimeHistogram = metrics.replicationControlResponseSendTimeInMs;
+          requestTotalTimeHistogram = metrics.replicationControlRequestTotalTimeInMs;
+          ReplicationControlAdminRequest replControlRequest =
+              ReplicationControlAdminRequest.readFrom(requestStream, adminRequest);
+          List<? extends PartitionId> partitionIds;
+          if (replControlRequest.getPartitionId() != null) {
+            error = validateRequest(replControlRequest.getPartitionId(), RequestOrResponseType.AdminRequest);
+            partitionIds = Collections.singletonList(replControlRequest.getPartitionId());
+          } else {
+            partitionIds = clusterMap.getAllPartitionIds();
+          }
+          if (!error.equals(ServerErrorCode.Partition_Unknown) && replicationManager.controlReplicationForPartitions(
+              partitionIds, replControlRequest.getOrigins(), replControlRequest.shouldEnable())) {
+            error = ServerErrorCode.No_Error;
+          } else {
+            logger.error("Could not set enable status for replication of {} from {} to {}. Origins list is empty or "
+                    + "contains an unknown datacenter", partitionIds, replControlRequest.getOrigins(),
+                replControlRequest.shouldEnable());
+            error = error == ServerErrorCode.No_Error ? ServerErrorCode.Unknown_Error : error;
```

Comment:
`validateRequest()` returns a `ServerErrorCode`. If that has changed the error code we want to retain it. But this does look confusing and the messaging in the error log needs to be updated. Will try to make this clearer